pkgs/http_profile/lib/http_profile.dart

```diff
@@ -2,8 +2,186 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:async' show StreamController, StreamSink;
+import 'dart:developer'
+    show addHttpClientProfilingData, getHttpClientProfilingData, Timeline;
 import 'dart:io';
 
+/// Describes an event that occurred while issuing an HTTP request.
+class HttpProfileRequestEvent {
+  /// The time at which the event occurred, in microseconds.
+  final int timestamp;
+  final String name;
+  final Map<String, dynamic>? arguments;
+
+  HttpProfileRequestEvent(this.timestamp, this.name, this.arguments);
+
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        'timestamp': timestamp,
+        'event': name,
+        if (arguments != null) 'arguments': arguments,
+      };
+}
+
+/// Describes proxy authentication details associated with an HTTP request.
+class HttpProfileProxyData {
+  final String? host;
+  final String? username;
+  final bool? isDirect;
+  final int? port;
+
+  HttpProfileProxyData({this.host, this.username, this.isDirect, this.port});
+
+  Map<String, dynamic> toJson() => <String, dynamic>{
+        if (host != null) 'host': host,
+        if (username != null) 'username': username,
+        if (isDirect != null) 'isDirect': isDirect,
+        if (port != null) 'port': port,
+      };
+}
+
+/// Describes details about an HTTP request.
+class HttpProfileRequestData {
+  Map<String, dynamic>? connectionInfo;
+
+  /// The content length of the request, in bytes.
+  int? contentLength;
+
+  /// The cookies presented to the server (in the 'cookie' header).
+  List<String>? cookies;
+
+  /// Events that occurred while issuing the request.
+  List<HttpProfileRequestEvent>? events;
```

Comment:
I should have made these fields private based on the way I envisioned this class to be used. I envisioned that a user of this package would have to populate an events array they defined themself and then use that to instantiate an `HttpProfileRequestData` object. Is it ok if I make these fields private?

`events` is currently under `HttpProfileRequestData` in the service extension response (https://github.com/dart-lang/sdk/blob/385c0063af8fbe77aff8ace14afb2a1c633dbdef/runtime/vm/service/service_extension.md?plain=1#L313-L316). It seems important to reflect that events can happen at any time, so I think we need to move the property up a level in the service response type. What do you think @bkonyi?

Reply:
> I should have made these fields private based on the way I envisioned this class to be used. I envisioned that a user of this package would have to populate an events array they defined themself and then use that to instantiate an `HttpProfileRequestData` object. Is it ok if I make these fields private?

It may be more convenient for some clients to provide information at different times than `HttpClient`. For example, neither `cupertino_http` nor `cronet_http` keep track of redirects so it would be easiest for them to add the redirect info to the profiling information as they go (i.e. somewhere in [this function](https://github.com/dart-lang/http/blob/f0a02f98f7c921e86ecc81c70f38bb6fbccc81b9/pkgs/cupertino_http/lib/src/cupertino_client.dart#L177) rather than store that information until they can construct an `HttpProfileRequestData`.

It also might be better UX in DevTools if you can track redirects even if the request fails.

> `events` is currently under `HttpProfileRequestData` in the service extension response (https://github.com/dart-lang/sdk/blob/385c0063af8fbe77aff8ace14afb2a1c633dbdef/runtime/vm/service/service_extension.md?plain=1#L313-L316). It seems important to reflect that events can happen at any time, so I think we need to move the property up a level in the service response type. What do you think @bkonyi?

You can change the location is this package without changing the service extension.

Reply:
> It seems important to reflect that events can happen at any time, so I think we need to move the property up a level in the service response type. What do you think @bkonyi?

I don't see any issue moving `events` up into `HttpProfileRequest` and would be okay with updating the protocol to match. This would be something relatively easy to work around in DevTools (check for `events` under `HttpProfileRequest` first and then check under `HttpProfileRequestData` if it's not there).

Reply:
> rather than store that information until they can construct an `HttpProfileRequestData`

By making the `HttpClientRequestProfile.requestData` setter only an entire object, I was trying to enforce the service response not having a populated `HttpProfileRequest.request` field until the initial request is completed (https://github.com/dart-lang/sdk/blob/385c0063af8fbe77aff8ace14afb2a1c633dbdef/runtime/vm/service/service_extension.md?plain=1#L271). We can allow users to populate fields of `HttpClientRequestProfile.requestData` as they go instead, but then we would need some other way of determining when the service response should have a populated `HttpProfileRequest.request` field. Some ideas are populating the field whenever `HttpProfileResponseData.endTime` is non-null, or populating the field whenever at least one field of `HttpClientRequestProfile.requestData` is non-null.

> You can change the location is this package without changing the service extension.

Yeah I just thought it might be confusing if we started putting events in there that weren't related to the initial request, since the service extension documentation says that the data in `HttpProfileRequest.request` is related to the initial request (https://github.com/dart-lang/sdk/blob/385c0063af8fbe77aff8ace14afb2a1c633dbdef/runtime/vm/service/service_extension.md?plain=1#L269). But, it looks like we're already doing that (https://github.com/dart-lang/sdk/blob/39e60fa3fc0820628ce63abc4adcaa2f9e048cc2/sdk/lib/_http/http_impl.dart#L203). So yeah, we don't have to move the property location in the service extension response type, but since @bkonyi's ok with it, we might as well.

Reply:
I think that giving maximum flexibility to package developers in how their populate fields is desirable.

```
  // Will not be provided if the initial request has not yet completed.
  HttpProfileRequestData request [optional];
```

Is the comment supposed to be descriptive or prescriptive, i.e., the receiver may not receive `request` if the request has not completed **or** the sender must not set `request` until the request has been completed? I suspect that the first interpretation is correct but I'm not sure.

Reply:
We have discussed this on a call and decided to allow users of this package to populate fields as they go, and to have the service extension RPCs always return all information that is available (i.e. we will not prevent information from being sent when a certain specific field has not yet been populated).